File: src/spaceObjects/planet.cpp
/*TODO

REGISTER_MULTIPLAYER_CLASS(Planet, "Planet");
Planet::Planet()
: SpaceObject(5000, "Planet")
{
    planet_size = 5000;
    cloud_size = 5200;
    planet_texture = "";
    cloud_texture = "";
    atmosphere_texture = "";
    atmosphere_size = 0;
    distance_from_movement_plane = 0;
    axial_rotation_time = 0.0;
    orbit_target_id = -1;
    orbit_time = 0.0f;
    orbit_distance = 0.0f;

    collision_size = -2.0f;

    setRadarSignatureInfo(0.5f, 0.f, 0.3f);

    addComponent<PlanetRenderer>();
    registerMemberReplication(&planet_size);
    registerMemberReplication(&cloud_size);
    registerMemberReplication(&atmosphere_size);
    registerMemberReplication(&planet_texture);
    registerMemberReplication(&cloud_texture);
    registerMemberReplication(&atmosphere_texture);
    registerMemberReplication(&atmosphere_color);
    registerMemberReplication(&distance_from_movement_plane);
    registerMemberReplication(&axial_rotation_time);
    addComponent<Orbit>();
    registerMemberReplication(&orbit_target_id);
    registerMemberReplication(&orbit_time);
    registerMemberReplication(&orbit_distance);

    if (entity) {
        entity.getOrAddComponent<NeverRadarBlocked>();
    }
}

void Planet::setPlanetAtmosphereColor(float r, float g, float b)
{
    atmosphere_color = glm::vec3{ r, g, b };
}

void Planet::setPlanetAtmosphereTexture(std::string_view texture_name)
{
    atmosphere_texture = texture_name;
}

void Planet::setPlanetSurfaceTexture(std::string_view texture_name)
{
    planet_texture = texture_name;
}

void Planet::setPlanetCloudTexture(std::string_view texture_name)
{
    cloud_texture = texture_name;
}

float Planet::getPlanetRadius()
{
    return planet_size;
}

float Planet::getCollisionSize()
{
    return collision_size;
}

void Planet::setPlanetRadius(float size)
{
    this->planet_size = size;
    this->cloud_size = size * 1.05f;
    this->atmosphere_size = size * 1.2f;
}

void Planet::setPlanetCloudRadius(float size)
{
    cloud_size = size;
}

void Planet::setDistanceFromMovementPlane(float distance_from_movement_plane)
{
    this->distance_from_movement_plane = distance_from_movement_plane;
}

void Planet::setAxialRotationTime(float time)
{
    axial_rotation_time = time;
}

void Planet::setOrbit(P<SpaceObject> target, float orbit_time)
{
    if (!target)
        return;
    this->orbit_target_id = target->getMultiplayerId();
    this->orbit_distance = glm::length(getPosition() - target->getPosition());
    this->orbit_time = orbit_time;
}

void Planet::update(float delta)
{
    if (collision_size == -2.0f)
    {
        updateCollisionSize();
        if (collision_size > 0.0f) {
            entity.getOrAddComponent<AvoidObject>().range = collision_size;
        }
    }

    if (orbit_distance > 0.0f)
    {
        P<SpaceObject> orbit_target;
        if (game_server)
            orbit_target = game_server->getObjectById(orbit_target_id);
        else
            orbit_target = game_client->getObjectById(orbit_target_id);
        if (orbit_target)
        {
            float angle = vec2ToAngle(getPosition() - orbit_target->getPosition());
            angle += delta / orbit_time * 360.0f;
            setPosition(orbit_target->getPosition() + vec2FromAngle(angle) * orbit_distance);
        }
    }

    if (axial_rotation_time != 0.0f)
        setRotation(getRotation() + delta / axial_rotation_time * 360.0f);
}

void Planet::drawOnRadar(sp::RenderTarget& renderer, glm::vec2 position, float scale, float rotation, bool long_range)
{
    if (collision_size > 0)
    {
        renderer.fillCircle(position, collision_size * scale, glm::u8vec4(atmosphere_color * 255.f, 128));
    }
}

void Planet::drawOnGMRadar(sp::RenderTarget& renderer, glm::vec2 position, float scale, float rotation, bool long_range)
{
    renderer.drawCircleOutline(position, planet_size * scale, 3, glm::u8vec4(255, 255, 255, 128));
}

void Planet::updateCollisionSize()
{
    if (std::abs(distance_from_movement_plane) >= planet_size)
    {
        collision_size = -1.0;
        entity.removeComponent<sp::Physics>();
    }else{
        collision_size = sqrt((planet_size * planet_size) - (distance_from_movement_plane * distance_from_movement_plane)) * 1.1f;
        entity.getOrAddComponent<sp::Physics>().setCircle(sp::Physics::Type::Static, collision_size);
    }
}
*/

File: scripts/api/entity/planet.lua
--- A Planet is a spherical piece of space terrain that can orbit other SpaceObjects.
--- Each Planet has separate textures for its surface, atmosphere, and cloud layers.
--- Several planetary textures are included in the resources/planets/ directory.
--- Planets can collide with objects and run callback functions upon collisions.
--- Examples:
--- -- Creates a small planetary system with a sun, a planet orbiting the sun, and a moon orbiting the planet.
--- sun = Planet():setPosition(5000, 15000):setPlanetRadius(1000):setPlanetAtmosphereTexture("planets/star-1.png"):setPlanetAtmosphereColor(1.0, 1.0, 1.0)
--- planet = Planet():setPosition(5000, 5000):setPlanetRadius(3000):setPlanetSurfaceTexture("planets/planet-1.png")
--- planet:setPlanetCloudTexture("planets/clouds-1.png"):setPlanetAtmosphereTexture("planets/atmosphere.png"):setPlanetAtmosphereColor(0.2, 0.2, 1.0):setOrbit(sun,40)
--- moon = Planet():setPosition(5000, 0):setPlanetRadius(1000):setPlanetSurfaceTexture("planets/moon-1.png"):setAxialRotationTime(20.0):setOrbit(planet,20)
--- @type creation
function Planet()
    local e = createEntity()
    e.components = {
        transform = {rotation=random(0, 360)},
        radar_signature = {gravity=0.5, biological=0.3},
        planet_render = {
            size=5000,
            cloud_size = 5200,
        },
        physics = {type="static", size=5000},
        never_radar_blocked = {}
    }
    return e
end

local Entity = getLuaEntityFunctionTable()
--- Sets this Planet's atmospheric effect color.
--- Example: planet:setPlanetAtmosphereColor(0.2,0.2,1.0) -- sets a blue atmosphere
function Entity:setPlanetAtmosphereColor(r, g, b)
    if self.components.planet_render then self.components.planet_render.atmosphere_color = {r, g, b} end
    return self
end
--- Sets this Planet's atmospheric effect texture.
--- Valid values are filenames of PNG files relative to the resources/ directory.
--- Optional; if defined, atmosphere textures should be transparent or translucent.
--- For stars, you can set an atmosphere texture such as planets/star-1.png with no surface texture.
--- Example: planet:setPlanetSurfaceTexture("planets/atmosphere.png")
function Entity:setPlanetAtmosphereTexture(texture)
    if self.components.planet_render then self.components.planet_render.atmosphere_texture = texture end
    return self
end
--- Sets this Planet's surface texture.
--- Valid values are filenames of PNG files relative to the resources/ directory.
--- Optional; if defined, surface textures should be opaque and use a 2:1-ratio equirectangular projection.
--- Example: planet:setPlanetSurfaceTexture("planets/planet-1.png")
function Entity:setPlanetSurfaceTexture(texture)
    if self.components.planet_render then self.components.planet_render.texture = texture end
    return self
end
--- Sets this Planet's cloud layer effect texture, which rotates independently of the planet.
--- Valid values are filenames of PNG files relative to the resources/ directory.
--- Optional; if defined, cloud layer textures should be transparent or translucent.
--- Example: planet:setPlanetCloudTexture("planets/cloud-1.png")
function Entity:setPlanetCloudTexture(texture)
    if self.components.planet_render then self.components.planet_render.cloud_texture = texture end
    return self
end
--- Returns this Planet's radius.
--- Example: planet:getPlanetRadius()
function Entity:getPlanetRadius()
    if self.components.planet_render then
        return self.components.planet_render.size
    end
    return 1000.0
end
--- Sets this Planet's radius, which also sets:
--- - its cloud radius to 1.05x this value
--- - its atmosphere radius to 1.2x this value
--- - its collision size to a function of this value and the planet's z-position
--- Defaults to 5000 (5U).
--- Example: planet:setPlanetRadius(2000)
function Entity:setPlanetRadius(size)
    local pr = self.components.planet_render
    if pr then
        pr.size = size
        pr.cloud_size = size*1.05
        pr.atmosphere_size = size*1.2
        if (pr.size * pr.size) > (pr.distance_from_movement_plane * pr.distance_from_movement_plane) then
            local collision_size = math.sqrt((pr.size * pr.size) - (pr.distance_from_movement_plane * pr.distance_from_movement_plane)) * 1.1;
            self.components.physics = {type="static", size=collision_size}
        else
            self.components.physics = nil
        end
    end
    return self
end
--- Sets this Planet's collision radius.
--- Defaults to a function of the Planet's radius and its z-position.
--- AI behaviors use this size to plot routes that try to avoid colliding with this Planet.
--- Example: planet:getCollisionSize()
function Entity:getCollisionSize()
    if self.components.physics then
        return self.components.physics.size
    end
    return 0.0
end
--- Sets this Planet's cloud radius, overriding Planet:setPlanetRadius().
--- Defaults to 1.05x this Planet's radius.
--- If this value isn't larger than the Planet's radius, the cloud layer won't be visible.
--- Example: planet:setPlanetCloudRadius(2500) -- sets this Planet's cloud radius to 2.5U
function Entity:setPlanetCloudRadius(radius)
    if self.components.planet_render then self.components.planet_render.cloud_size = radius end
    return self
end
--- Sets the z-position of this Planet, the distance by which it's offset above (positive) or below (negative) the movement plane.
--- This value also modifies the Planet's collision radius.
--- Defaults to 0.
--- Example: planet:setDistanceFromMovementPlane(-500) -- sets the planet 0.5U below the movement plane
function Entity:setDistanceFromMovementPlane(z)
    local pr = self.components.planet_render
    if pr then
        pr.distance_from_movement_plane = z
        if (pr.size * pr.size) > (pr.distance_from_movement_plane * pr.distance_from_movement_plane) then
            local collision_size = math.sqrt((pr.size * pr.size) - (pr.distance_from_movement_plane * pr.distance_from_movement_plane)) * 1.1;
            self.components.physics = {type="static", size=collision_size}
        else
            self.components.physics = nil
        end
    end
    return self
end
--- Sets this Planet's axial rotation time, seconds per full rotation.
--- Defaults to 0.
--- Example: planet:setAxialRotationTime(20)
function Entity:setAxialRotationTime(rotation_time)
    if rotation_time ~= 0.0 then
        self.components.spin = {rate=360.0/rotation_time}
    else
        self.components.spin = nil
    end
    return self
end
--- Sets a SpaceObject around which this SpaceObject orbits, as well as its orbital period in seconds. Setting time to 0 will stop movement, while still being locked in orbit.
--- An orbit can be cancelled by setting the target to nil
--- Example:
--- moon:setOrbit(planet, 20)
--- moon:setOrbit(nil) -- undo orbiting
function Entity:setOrbit(target, time)
    if target == nil then
        self.components.orbit = nil
        return self
    end
    local x0, y0 = self:getPosition()
    local x1, y1 = target:getPosition()
    local xd, yd = (x1 - x0), (y1 - y0)
    local distance = math.sqrt(xd * xd + yd * yd)
    self.components.orbit = {
        target = target,
        center = {x1, y1},
        distance = distance,
        time = time,
    }
    return self
end
